astroplan/constraints.py

```diff
@@ -0,0 +1,577 @@
+        upper bound one attosecond later than the input time.
+
+    time_resolution : `~astropy.units.quantity` (optional)
+        Time-grid spacing
+
+    Returns
+    -------
+    times : `~astropy.time.Time`
+        Linearly-spaced sequence of times
+    """
+    if time_range.isscalar:
+        return Time([time_range])
+    else:
+        return Time(np.arange(time_range[0].jd, time_range[1].jd,
+                              time_resolution.to(u.day).value), format='jd')
+
+@abstractmethod
+class Constraint(object):
+    """
+    Abstract class for objects defining observational constraints.
+    """
+    __metaclass__ = ABCMeta
+
+    def __call__(self, time_range, observer, targets,
+                 time_resolution=DEFAULT_TIME_RESOLUTION):
+
+        if not isinstance(time_range, Time):
+            time_range = Time(time_range)
+        cons = self._compute_constraint(time_range, observer, targets,
+                                        time_resolution=time_resolution)
+        return cons
+
+    def _compute_constraint(self, time_range, observer, targets):
+        # Should be implemented on each subclass of Constraint
+        raise NotImplementedError
+
+    def _get_altaz(self, time_range, observer, targets,
+                   time_resolution=DEFAULT_TIME_RESOLUTION,
+                   force_zero_pressure=False):
+        """
+        Calculate alt/az for ``target`` at times linearly spaced between
+        the two times in ``time_range`` with grid spacing ``time_resolution``
+        for ``observer``.
+
+        Cache the result on the ``observer`` object.
+
+        Parameters
+        ----------
+        time_range : `~astropy.time.Time`
+            Lower and upper time bounds on which to compute the constraints.
+            If input is a scalar time, will only compute at that time (no
+            range).
+
+        targets : {list, `~astropy.coordinates.SkyCoord`, `~astroplan.FixedTarget`}
+            Target or list of targets
+
+        observer : `~astroplan.Observer`
+            The observer who has constraints ``constraints``
+
+        time_resolution : `~astropy.units.Quantity` (optional)
+            Set the time resolution in calculations of the altitude/azimuth
+
+        Returns
+        -------
+        altaz_dict : dict
+            Dictionary containing two key-value pairs. (1) 'times' contains the
+            times for the alt/az computations, (2) 'altaz' contains the
+            corresponding alt/az coordinates at those times.
+        """
+        if not hasattr(observer, '_altaz_cache'):
+            observer._altaz_cache = {}
+
+        times = time_grid_from_range(time_range,
+                                     time_resolution=time_resolution)
+
+        # convert times, targets to tuple for hashing
+        aakey = (tuple(times.jd), tuple(targets))
+
+        if aakey not in observer._altaz_cache:
+            if force_zero_pressure:
+                observer_old_pressure = observer.pressure
```

Comment:
You can use a `try...finally` to avoid the problem of the `observer` having the `pressure` not restored because of e.g. throwing an exception in `altaz`.  You would do this by:
```python
try:
    if force_zero_pressure:
        observer_old_pressure = observer.pressure
        observer.pressure = 0

    altaz = observer.altaz(times, targets)
   observer._altaz_cache[aakey] = dict(times=times,
                                                              altaz=altaz)
finally:
    if force_zero_pressure:
        observer.pressure = observer_old_pressure
```